src/server/engine_shard_set.cc

```diff
@@ -39,8 +52,45 @@ using absl::GetFlag;
 
 namespace {
 
+constexpr DbIndex DEFAULT_DB_INDEX = 0;
+
 vector<EngineShardSet::CachedStats> cached_stats;  // initialized in EngineShardSet::Init
 
+inline bool MemUsageChanged(uint64_t current, uint64_t now) {
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+  // we only care whether the memory usage is going up
+  // otherwise we don't really case
+  return (current * commit_use_threshold) < float(now);
+}
+
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool IsMemDefragRequired(uint64_t* current_commit, uint64_t* current_use) {
+  static const uint64_t mem_size = max_memory_limit;
+  static const uint64_t threshold_mem = mem_size / GetFlag(FLAGS_mem_defrag_threshold);
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  // we want to make sure that we are not running this to many times - i.e.
+  // if there was no change to the memory, don't run this
+  if (MemUsageChanged(*current_commit, commited) || MemUsageChanged(*current_use, mem_in_use)) {
```

Comment:
I prefer to omit MemUsageChanged for this PR.

Reply:
I will make it no op for now